### DISCUSSION TITLE
Real USB FDD

### DISCUSSION THREAD

**Discussion by drrak** [86Box/86Box#3343] [Ideas]
_Translator:_
It would be great if the program detects a real external USB FDD.

**Comment by Cacodemon345** [comment]
We don't even have proper USB emulation, let alone USB passthrough emulation.

**Comment by drrak** [comment]
Then reading non-IMG files would be useful. You could set it to look for files in USB-FDD.
It is possible?

**Comment by FavoritoHJS** [comment]
I'm guessing he means "if you plug in an external floppy drive, you can configure a machine to use it".

Which is drive passthrough with extra steps, so Eventually:tm:

**Comment by FavoritoHJS** [comment]
> Then reading non-IMG files would be useful. You could set it to look for files in USB-FDD. It is possible?

You could image the floppy disks you will use...

**Comment by dhrdlicka** [comment]
> Then reading non-IMG files would be useful. You could set it to look for files in USB-FDD. It is possible?

There are several other formats already supported for floppy images aside from IMG.

Either way I am not quite sure if drive passthrough would be happening for USB floppy drives, since they don't tend to allow direct access to the floppy drive controller.

**Comment by drrak** [comment]
I'm not asking how to prepare or run IMG, I only care about loading floppy disks via USB-FDD.

**Comment by Torinde** [comment]
USB passthrough: #2907

**Comment by fmillion** [comment]
Sounds like it's essentially a case of block device mapping. From a Linux perspective, you want to make a block device (e.g. `/dev/fd0` or `/dev/sda` for a USB floppy) available to the VM as if it's a block device attached to the VM. On Windows you can accomplish something similar with a device like `\\.\PHYSICALDRIVE1`. but you have to run as administrator to get that access. (on Linux you can just add your user to the `disk` group, or I think you can get even more granular with cgroups)

I'd imagine something like that wouldn't be _too_ difficult, but it wouldn't be _easy_ either. In both the floppy and the hard drive emulation layer you would need to basically support opening a physical block device as well as an image file. More importantly, you would need to be able to infer the geometry from the size of the block device, which would involve extra syscalls (since block device files don't have a "size" per se like image files do). You would also need to handle the disk-change flag so that the virtual machine sees disk changes and capacity changes (e.g. putting a DD disk in an HD drive). 

There could be some usefulness here beyond floppies. You could map a real IDE hard drive attached to a USB adapter to a VM and boot an old real Win98 install for example. Or you could attach a real USB Zip drive and access your real physical Zip disks in the VM directly. Or it could allow you to use a real optical drive.

I'd say this is a "nice-to-have" since there's plenty of ways to image disks, CDs, Zip disks, etc. out there if you have working hardware. Nonetheless it would be cool if you have a huge collection of old floppies and want to just "use" them as-is.
